HDDS-5461. Move old objects to delete table on overwrite

kuenishi opened the pull request:
## What changes were proposed in this pull request?

HDDS-5243 was a patch for omitting key locations for clients on reading. But the same warning of large response size observed in our cluster for putting data. This is harmful in terms of retry storm, as hadoop-rpc handles this large-response-exception as retry-able exception. Thus, RetryInvocationHandler retries, despite it cannot be recovered by retry, for 15 times, receiving large response message exceeding default limit of RPC message size 128MB as follows.

```
    2021-06-21 19:23:10,717 [IPC Server handler 65 on default port 9862] WARN org.apache.hadoop.ipc.Server: Large response size 134538349 for call Call#2037538 Retry#15 org.apache.hadoop.ozone.om.protocol.OzoneManagerProtocol.submitRequest from 10.192.17.172:34070
    2021-06-21 19:23:10,722 [IPC Server handler 65 on default port 9862] WARN org.apache.hadoop.ipc.Server: IPC Server handler 65 on default port 9862, call Call#2037538 Retry#15 org.apache.hadoop.ozone.om.protocol.OzoneManagerProtocol.submitRequest from 10.192.17.172:34070: output error
    2021-06-21 19:23:10,722 [IPC Server handler 65 on default port 9862] INFO org.apache.hadoop.ipc.Server: IPC Server handler 65 on default port 9862 caught an exception
    java.nio.channels.AsynchronousCloseException
    at java.nio.channels.spi.AbstractInterruptibleChannel.end(AbstractInterruptibleChannel.java:205)
    at sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:478)
    at org.apache.hadoop.ipc.Server.channelIO(Server.java:3642)
    at org.apache.hadoop.ipc.Server.channelWrite(Server.java:3594)
    at org.apache.hadoop.ipc.Server.access$1700(Server.java:139)
    at org.apache.hadoop.ipc.Server$Responder.processResponse(Server.java:1657)
    at org.apache.hadoop.ipc.Server$Responder.doRespond(Server.java:1727)
```

Suggestion in HDDS-5393 was wrong and it shall be fixed by making old blocks eligible for deletion service, moving to deletion table. It is only needed for normal object-put, while not needed for MultipartUpload objects, if I understand correctly. 

## What is the link to the Apache JIRA

[HDDS-5461](https://issues.apache.org/jira/browse/HDDS-5461)

Please replace this section with the link to the Apache JIRA)

## How was this patch tested?

- Add tiny unit test that checks to `TestOMKeyCommitResponse`
- Manual & local test using AWS CLI with debug log

kuenishi commented:
Note: @bharatviswa504 has been also working on this concurrently in #2432 .

bharatviswa504 reviewed (commented):
If I understand correctly this PR does below, let me know if i am missing something here.

When key is override, in Key Create Request it generated Key(appended with session id) and generate keyInfo with only its blocks, but using the version number as oldVersion + 1. 
And finally in keyCommit, deleting the old Key.

I think, what we can do here to solve the issue and also with out removing versioning code.

1. Append the new blocks with newer version only if bucket versioning enabled. (Instead of previous blocks appending), and we can have your logic for versioning disabled buckets.
2. During keyCreate, based on latest version flag return latest blocks only (By default this config is true, so it works for all upgraded clusters) (MY other PR has taken care of this #2432 )

And also it needs similar update for File requests also (OMFileCreateRequest)

bharatviswa504 commented:
> Note: @bharatviswa504 has been also working on this concurrently in #2432 .

As mentioned in the Jira, my PR addresses during key create return only latest version of key. 

Key1 create - Blocks Version0 B1(0), B2(0), B3(0)

1st time override
Key1 override - Version0 -> B1(0), B2(0), B3(0), Version1 -> B1(0), B2(0), B3(0), New Blocks for this version (B4(1), B5(1), B6(1))

2nd time override
Key1 override - Version0 B1(0), B2(0), B3(0), Version1 -> B1(0), B2(0), B3(0), New Blocks for this version (B4(1), B5(1), B6(1)), Version 2 -> B1(0), B2(0), B3(0), B4(1), B5(1), B6(1), B7(2), B8(2), B9(2)

So, the problem is each time a key is override, each version has its version of blocks along with all older versions.

So, now when key create it returns all this versions. So, the list can become huge in size, depending on number of times key is override. In this patch I will fix to return latest version for create key. 

>(I will open a new Jira to discuss about having each version including all its previous version)

This PR is proposing to solve the 2nd part of the problem.

kuenishi commented:
OK, I agree with your idea of preserving versioning code. I'll add some work on this, but, AFAIK, such configuration to switch versioning on/off doesn't exist yet. We have to add such property in ozone-site.xml. I want to discuss the name of property before starting my work. It will be hidden by default, but it won't be good idea to change the name afterwards.

Concurrently I'll take a look on OmFileCreateRequest and prepare update.

Also, I have another question on `KeyManager` - it isn't used any more, but should it be updated and tested as well?

bharatviswa504 commented:
> OK, I agree with your idea of preserving versioning code. I'll add some work on this, but, AFAIK, such configuration to switch versioning on/off doesn't exist yet. We have to add such property in ozone-site.xml. I want to discuss the name of property before starting my work. It will be hidden by default, but it won't be good idea to change the name afterwards.
> 
> Concurrently I'll take a look on OmFileCreateRequest and prepare update.
> 
> Also, I have another question on `KeyManager` - it isn't used any more, but should it be updated and tested as well?

We have isVersionEnabled flag already, no where we use it on the server. Do you think still we need to have config for this?
message BucketInfo {
    required string volumeName = 1;
    required string bucketName = 2;
    repeated OzoneAclInfo acls = 3;
    required bool isVersionEnabled = 4 [default = false];

kuenishi commented:
> Do you think still we need to have config for this?

No, we don't. Thank you for the guidance.

kuenishi commented:
> And also it needs similar update for File requests also (OMFileCreateRequest)

As the code update is all in  `OMKeyRequest#prepareFileInfo()`, OMFileCreateRequest shares code of preparing OmKeyInfo with OMKeyCreateRequest and OMkeyCreateRequestWithFSO.

ChenSammi commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmKeyLocationInfoGroup.java:
```diff
@@ -143,16 +143,22 @@ public static OmKeyLocationInfoGroup getFromProtobuf(
+      List<OmKeyLocationInfo> newLocationList, boolean keepVersions) {
+    Map<Long, List<OmKeyLocationInfo>> newMap = null;
+    if (keepVersions) {
+      newMap = new HashMap<>(locationVersionMap);
```
We don‘t need to include the locationVersionMap to the new  OmKeyLocationInfoGroup even when version is true.  
The current OmKeyLocationInfoGroup design actually has one point to improve.   You can refer to @symious document https://docs.google.com/document/d/1UYVETpSHyVppNPjHq8VuXj2t2Y0LYQNRYa6awSTCROA/edit#heading=h.h2k2ovrsrw33 for more detail information.

kuenishi replied:
Thank you for the review. I updated the commented  part so that it does keep past location version map and also added option 3 of my idea in this issue to the @symious document.

I'm sorry that I don't have much time to add necessary unit tests and make integration tests passing for now, but will try in a few days.

kuenishi commented:
@ChenSammi @bharatviswa504 I have added option 3 to @symious 's [document](https://docs.google.com/document/d/1UYVETpSHyVppNPjHq8VuXj2t2Y0LYQNRYa6awSTCROA/edit#heading=h.h2k2ovrsrw33) and implemented it in this pull request. Multipart-uploaded objects shall be done in HDDS-5656 to keep this review smaller. PTAL!

vivekratnavel commented:
@bharatviswa504 @ChenSammi Can you please take another look at this patch and help merge this?

bharatviswa504 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequestWithFSO.java:
```diff
@@ -143,6 +149,22 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
+      }
+
       // Add to cache of open key table and key table.
       OMFileRequest.addOpenFileTableCacheEntry(omMetadataManager, dbFileKey,
```
```
      if (keysToDelete != null) {
        omMetadataManager.getDeletedTable().addCacheEntry(
            new CacheKey<>(dbOzoneKey),
            new CacheValue<>(Optional.of(keysToDelete), trxnLogIndex));
      }
```

Something like this

bharatviswa504 reviewed (changes requested):
Sorry for the delay.
I have few comments.
And also have added a test for my understanding purpose to test, feel free to use it which covers both versioning and with out versioning.

https://github.com/bharatviswa504/hadoop-ozone/commit/b0d9df21ea9fec8365a2da45451db44d881213a2

bharatviswa504 commented:
@kuenishi 
Have you got a chance to look at latest comments. Let me know if any help is needed?

kuenishi commented:
I think I've addressed all comments, but I'll need a bit more hustle to pass the tests.

bharatviswa504 commented:
@kuenishi I was busy with some other tasks, will get to this today,
